src/eos-updater-apply.c

```diff
@@ -73,6 +74,431 @@ get_test_osname (void)
+{
+  g_autofree gchar *temp_dir = g_dir_make_tmp ("ostree-checkout-XXXXXX", error);
+
+  if (!temp_dir)
+    return NULL;
+
+  return g_build_filename (temp_dir, "checkout", NULL);
+}
+
+static gboolean
+parse_flatpak_ref_kind (const gchar *string, FlatpakRefKind *kind)
+{
+  g_return_val_if_fail (kind != NULL, FALSE);
+
+  if (g_strcmp0 (string, "app") == 0)
+    {
+      *kind = FLATPAK_REF_KIND_APP;
+      return TRUE;
+    }
+
+  if (g_strcmp0 (string, "runtime") == 0)
+    {
+      *kind = FLATPAK_REF_KIND_RUNTIME;
+      return TRUE;
+    }
+
+  return FALSE;
+}
+
+static const gchar *
+format_flatpak_ref_kind (FlatpakRefKind kind)
+{
+  switch (kind)
+    {
+      case FLATPAK_REF_KIND_APP:
+        return "app";
+      case FLATPAK_REF_KIND_RUNTIME:
+        return "runtime";
+      default:
+        return NULL;
+    }
+}
+
+static GPtrArray *
+flatpaks_to_deploy_from_commit (OstreeRepo   *repo,
+                                const gchar  *checksum,
+                                GCancellable *cancellable,
+                                GError       **error)
+{
+  g_autofree gchar *checkout_directory = get_temporary_directory_to_checkout_in (error);
+  g_autoptr(GFile) checkout_directory_path = NULL;
+  OstreeRepoCheckoutAtOptions options = { 0, };
+  g_autofree gchar *autoinstall_flatpak_contents = NULL;
+  g_autofree gchar *autoinstall_flatpak_file_path = NULL;
+  g_autoptr(GRegex) autoinstall_flatpak_regex = NULL;
+  g_auto(GStrv) autoinstall_flatpak_lines = NULL;
+  GPtrArray *autoinstall_flatpaks = NULL;
+  GStrv autoinstall_flatpak_lines_iter = NULL;
+
+  if (!checkout_directory)
+    return FALSE;
+
+  checkout_directory_path = g_file_new_for_path (checkout_directory);
+
+  autoinstall_flatpak_regex = g_regex_new ("(.*),(.*),(.*)", 0, 0, error);
+  if (!autoinstall_flatpak_regex)
+    return NULL;
+
+  /* Now that we have a temporary directory, checkout the OSTree in it
+   * at the /usr/share/eos-updater path. If it fails, there's nothing to
+   * read, otherwise we can read in the list of flatpaks to be auto-installed
+   * for this commit. */
+  options.subpath = "usr/share/eos-updater";
+
+  if (!ostree_repo_checkout_at (repo,
+                                &options,
+                                -1,
+                                checkout_directory,
+                                checksum,
+                                cancellable,
+                                error))
+    {
+      rmdir (checkout_directory);
+      return NULL;
+    }
+
+  autoinstall_flatpak_file_path = g_build_filename (checkout_directory,
+                                                    "flatpak.autoinstall",
+                                                    NULL);
+
+  /* Repository checked out, read usr/share/eos-updater/flatpak.autoinstall
+   * to determine which flatpaks should be auto-installed. Recall that this
+   * takes the format of repo,com.bundle.Identifier/arch/branch */
```

Comment:
The conventional format would be `$remote_name:$flatpak_ref` (colon rather than comma). That format is what OSTree uses everywhere (it’s a ‘refspec’). The benefit of using refspecs is that you can then use `ostree_parse_refspec()` on them and save writing another parser.

Reply:
Being able to lean on `ostree_parse_refspec` would be very handy indeed.